src/uint/div.rs

```diff
@@ -660,6 +602,283 @@ impl<const LIMBS: usize> RemLimb for Uint<LIMBS> {
+        }
+        yc if yc > N1 => {
+            // Divisor is greater than dividend. Return zero and the dividend as the
+            // quotient and remainder
+            return (Uint::ZERO, dividend.resize());
+        }
+        _ => {}
+    };
+
+    let lshift = (Limb::BITS - (dbits % Limb::BITS)) % Limb::BITS;
+    let rshift = if lshift == 0 { 0 } else { Limb::BITS - lshift };
+    let mut x = dividend.to_limbs();
+    let mut x_hi = Limb::ZERO;
+    let mut xi = N1 - 1;
+    let mut y = divisor.to_limbs();
+    let mut i;
+    let mut carry;
+
+    if lshift != 0 {
+        // Shift divisor such that it has no leading zeros
+        // This means that div2by1 requires no extra shifts, and ensures that the high word >= b/2
+        i = 0;
+        carry = Limb::ZERO;
+        while i < yc {
+            (y[i], carry) = (Limb((y[i].0 << lshift) | carry.0), Limb(y[i].0 >> rshift));
+            i += 1;
+        }
+
+        // Shift the dividend to match
+        i = 0;
+        carry = Limb::ZERO;
+        while i < N1 {
+            (x[i], carry) = (Limb((x[i].0 << lshift) | carry.0), Limb(x[i].0 >> rshift));
+            i += 1;
+        }
+        x_hi = carry;
+    }
+
+    let reciprocal = Reciprocal::new(y[yc - 1].to_nz().expect("zero divisor"));
+
+    loop {
+        // Divide high dividend words by the high divisor word to estimate the quotient word
+        let (mut quo, mut rem) = div2by1(x_hi.0, x[xi].0, &reciprocal);
+
+        i = 0;
+        while i < 2 {
+            let qy = (quo as WideWord) * (y[yc - 2].0 as WideWord);
+            let rx = ((rem as WideWord) << Word::BITS) | (x[xi - 1].0 as WideWord);
+            // Constant-time check for q*y[-2] < r*x[-1], based on ConstChoice::from_word_lt
+            let diff = ConstChoice::from_word_lsb(
+                ((((!rx) & qy) | (((!rx) | qy) & (rx.wrapping_sub(qy)))) >> (WideWord::BITS - 1))
+                    as Word,
+            );
+            quo = quo.saturating_sub(diff.select_word(0, 1));
+            rem = rem.saturating_add(diff.select_word(0, y[yc - 1].0));
+            i += 1;
+        }
+
+        // Subtract q*divisor from the dividend
+        carry = Limb::ZERO;
+        let mut borrow = Limb::ZERO;
+        let mut tmp;
+        i = 0;
+        while i < yc {
+            (tmp, carry) = Limb::ZERO.mac(y[i], Limb(quo), carry);
+            (x[xi + i + 1 - yc], borrow) = x[xi + i + 1 - yc].sbb(tmp, borrow);
+            i += 1;
+        }
+        (_, borrow) = x_hi.sbb(carry, borrow);
+
+        // If the subtraction borrowed, then decrement q and add back the divisor
+        let ct_borrow = ConstChoice::from_word_mask(borrow.0);
+        carry = Limb::ZERO;
+        i = 0;
+        while i < yc {
+            (x[xi + i + 1 - yc], carry) =
+                x[xi + i + 1 - yc].adc(Limb::select(Limb::ZERO, y[i], ct_borrow), carry);
+            i += 1;
+        }
+        quo -= ct_borrow.select_word(0, 1);
```

Comment:
I can't actually get this to trigger in practice, removing it seems to have no effect as `quo` is always correct before this. It would be nice to prove that (mathematically).